prow/tide/tide.go

```diff
@@ -179,18 +180,90 @@ func byRepoAndNumber(prs []PullRequest) map[string]PullRequest {
 	return m
 }
 
+func requirementDiff(pr *PullRequest, q *config.TideQuery) (string, int) {
+	const maxLabelChars = 50
+	var desc string
+	// Drops labels if needed to fit the description text area, but keep at least 1.
+	truncate := func(labels []string) []string {
+		i := 1
+		chars := len(labels[0])
+		for ; i < len(labels); i++ {
+			if chars+len(labels[i]) > maxLabelChars {
+				break
+			}
+			chars += len(labels[i]) + 2 // ", "
+		}
+		return labels[:i]
+	}
+
+	var missingLabels []string
+	for _, l1 := range q.Labels {
+		var found bool
+		for _, l2 := range pr.Labels.Nodes {
+			if string(l2.Name) == l1 {
+				found = true
+				break
+			}
+		}
+		if !found {
+			missingLabels = append(missingLabels, l1)
+		}
+	}
+	if len(missingLabels) > 0 {
+		sort.Strings(missingLabels)
+		trunced := truncate(missingLabels)
+		if len(trunced) == 1 {
+			desc = fmt.Sprintf(" Needs %s label.", trunced[0])
+		} else {
+			desc = fmt.Sprintf(" Needs %s labels.", strings.Join(trunced, ", "))
+		}
+	}
+
+	var presentLabels []string
+	for _, l1 := range q.MissingLabels {
+		for _, l2 := range pr.Labels.Nodes {
+			if string(l2.Name) == l1 {
+				presentLabels = append(presentLabels, l1)
+				break
+			}
+		}
+	}
+	if desc == "" && len(presentLabels) > 0 {
+		sort.Strings(presentLabels)
+		trunced := truncate(presentLabels)
+		if len(trunced) == 1 {
+			desc = fmt.Sprintf(" Should not have %s label.", trunced[0])
+		} else {
+			desc = fmt.Sprintf(" Should not have %s labels.", strings.Join(trunced, ", "))
+		}
+	}
+
+	// TODO(cjwagner): List reviews (states:[APPROVED], first: 1) as part of open
+	// PR query and include status context description.
+
+	return desc, len(missingLabels) + len(presentLabels)
+}
+
 // Returns expected status state and description.
```

Comment:
Doc how it chooses the query your PR is closest to fulfilling?

Reply:
Done.